### PR TITLE
MRG: Actually do projection

### PR DESCRIPTION
**Pull request by Eric89GXL**
@agramfort it's probably good to take a look at this one now, esp. how `test_epochs.py` had to be changed. Writing out `EpochsArray` and reading it back by default will have avg ref added and applied.

The least intrusive change might actually be to have `add_eeg_ref=False` in `read_epochs`. If people have been using FIF, they've presumably already added the ref by that stage (during `Epochs` construction), and it avoids breaking people's `EpochsArray` use.

### CODE REVIEW

File: mne/epochs.py [hunk 1]
```diff
@@ -1913,7 +1920,8 @@ def __init__(self, data, info, events, tmin=0, event_id=None,
         super(EpochsArray, self).__init__(info, data, events, event_id, tmin,
                                           tmax, baseline, reject=reject,
                                           flat=flat, reject_tmin=reject_tmin,
-                                          reject_tmax=reject_tmax, decim=1)
+                                          reject_tmax=reject_tmax, decim=1,
+                                          add_eeg_ref=False)
```
Comment (Eric89GXL) [inline mne/epochs.py]: We were silently always adding an `EEG` ref before but not applying it (whoops). So we could consider this a bug fix...?

File: mne/tests/test_epochs.py [hunk 2]
```diff
@@ -1826,7 +1844,7 @@ def test_array_epochs():
     assert_allclose(epochs.times, [0.])
     assert_allclose(epochs.get_data(), data[:, :, :1])
     epochs.save(temp_fname)
-    epochs_read = read_epochs(temp_fname)
+    epochs_read = read_epochs(temp_fname, add_eeg_ref=False)
```
Comment (Eric89GXL) [inline mne/tests/test_epochs.py]: These are the changes that I suspect will break people's code. I could try changing `add_eeg_ref=False` by default and seeing what the result looks like, but then we'll have `proj=True, add_eeg_ref=False` which is the opposite of what you suggested @agramfort. This is a mess...

Comment (Eric89GXL) [inline mne/tests/test_epochs.py]: Confirmed that setting `read_epochs(..., avg_eeg_ref=False)` allows these `EpochsArray` changes to be reverted without breaking any other tests.

Comment (Eric89GXL) [inline mne/tests/test_epochs.py]: (setting that to the default of the function, that is)

File: mne/epochs.py (hunk 1 above)
Comment (agramfort) [inline mne/epochs.py]: if we clearly document it in the what's new fine with me.

File: doc/whats_new.rst [hunk 3]
```diff
@@ -58,6 +60,8 @@ API
 
     - :func:`mne.io.read_raw_brainvision` now has ``event_id`` argument to assign non-standard trigger events to a trigger value by `Teon Brooks`_
 
+    - :func:`mne.read_epochs` now has ``add_eeg_ref=False`` by default, since average EEG reference can be added before writing or after reading using the method :func:`mne.Epochs.add_eeg_ref`.
+
```
Comment (Eric89GXL) [inline doc/whats_new.rst]: @agramfort from looking at the code, it really seems like the cleanest solution is to eliminate the `add_eeg_ref` arguments to our constructors altogether. It's the most explicit about what's happening to data. Maybe target for 0.12 so we can test thoroughly?

File: mne/source_space.py [hunk 4]
```diff
@@ -229,50 +229,35 @@ def export_volume(self, fname, include_surfaces=True,
...
+            if mri_resolution:
+                # get the values for this volume
+                use = id_ * (aseg_data == id_).astype(int).ravel('F')
             else:
-                # update the inuse array
-                if mri_resolution:
-
-                    # get the values for this volume
-                    use = i * (aseg.get_data() == i).astype(int)
-                    inuse += use.ravel((2, 1, 0))
```
Comment (Eric89GXL) [inline mne/source_space.py]: This is a fix for deprecated use of indices in `ravel`, can split into a different PR if necessary, but both are needed for 0.11

Comment (Eric89GXL) [inline mne/source_space.py]: Ahh never mind I'll move it...

**Comment by Eric89GXL** [comment]
Ready for review/merge from my end. Once this is in, #2676 should work okay with `add_eeg_ref` options, whatever those may be. My vote has actually changed to not add the `add_eeg_ref` argument at all, since users with this PR can do `epochs.add_eeg_ref()` afterward if they want.

**Comment by agramfort** [comment]
@jasmainak how does this look to you?

**Comment by jasmainak** [comment]
Looks like we have failing tests. I check carefully tomorrow ..

**Comment by Eric89GXL** [comment]
Fixing now

**Comment by Eric89GXL** [comment]
@jasmainak let me know if this is good for you. This and your PR are the only things holding up release I think.

File: doc/whats_new.rst (hunk 3 above)
Comment (jasmainak) [inline doc/whats_new.rst]: shouldn't this actually go to the bug section? Otherwise, users will be confused why there is no deprecation, no?

**Comment by jasmainak** [comment]
I rebased this onto my PR and it works like a charm @Eric89GXL :)

File: doc/whats_new.rst (hunk 3 above)
Comment (Eric89GXL) [inline doc/whats_new.rst]: I'm not actually sure. @agramfort where do you prefere these messages to go?

Comment (jasmainak) [inline doc/whats_new.rst]: I agree what you say about eliminating `add_eeg_ref` from the constructor. We have too many arguments to the `Epochs` class anyway ...

File: mne/tests/test_epochs.py [hunk 5]
```diff
@@ -724,6 +727,29 @@ def test_epochs_proj():
...
+                    baseline=(None, 0), preload=True, add_eeg_ref=False)
+    epochs.pick_channels(['EEG 001', 'EEG 002'])
+    assert_equal(len(epochs), 7)  # sufficient for testing
+    temp_fname = 'test.fif'
+    epochs.save(temp_fname)
+    epochs = read_epochs(temp_fname, add_eeg_ref=True, proj=True)
+    assert_allclose(epochs.get_data().mean(axis=1), 0, atol=1e-15)
+    epochs = read_epochs(temp_fname, add_eeg_ref=True, proj=False)
+    assert_raises(AssertionError, assert_allclose,
+                  epochs.get_data().mean(axis=1), 0., atol=1e-15)
```
Comment (jasmainak) [inline mne/tests/test_epochs.py]: so, you shouldn't get an error here if `preload` was `False`?

File: mne/io/proj.py [hunk 6]
```diff
@@ -100,6 +100,13 @@ def add_proj(self, projs, remove_existing=False):
 
         return self
 
+    def add_eeg_ref(self):
+        """Add an average EEG reference projector if one does not exist
+        """
+        if _needs_eeg_average_ref_proj(self.info):
+            eeg_proj = make_eeg_average_ref_proj(self.info, activate=True)
```
Comment (jasmainak) [inline mne/io/proj.py]: What does the `activate` do actually? It's not super clear to me ...

File: mne/tests/test_epochs.py (hunk 5 above)
Comment (Eric89GXL) [inline mne/tests/test_epochs.py]: the `preload` state shouldn't matter

Comment (Eric89GXL) [inline mne/tests/test_epochs.py]: do you want me to test that case, too?

Comment (jasmainak) [inline mne/tests/test_epochs.py]: yeah maybe, I thought projs were applied during `get_data()` if preload was False ...

File: mne/io/proj.py (hunk 6 above)
Comment (Eric89GXL) [inline mne/io/proj.py]: Ahh -- I think this actually needs to be `False` now that I think about it. IIRC `activate` means set the `proj['active'] = activate`, and `proj['active']` says that the projector has been applied to the data already. @agramfort is that right?

File: mne/tests/test_epochs.py (hunk 5 above)
Comment (Eric89GXL) [inline mne/tests/test_epochs.py]: It should be that projs are applied during `get_data()` if `proj=True` regardless of preload state. Preload state should never change what you get from `get_data`, if it does there is a bug with our preloading logic. I'll add a test though to be sure.

File: mne/io/proj.py (hunk 6 above)
Comment (jasmainak) [inline mne/io/proj.py]: yeah, that's what I thought. Although, `activate` is slightly confusing to me. I would say `mark_active` or something like that. Because `activate` seems to give the impression that something was done to the data ... but that's a separate discussion altogether.

File: mne/tests/test_epochs.py (hunk 5 above)
Comment (jasmainak) [inline mne/tests/test_epochs.py]: So, if `preload=True` the proj is applied a second time during `get_data()`? I guess it doesn't matter? You can apply the same proj as many times, right?

Comment (jasmainak) [inline mne/tests/test_epochs.py]: Oh wait, I guess you check if the proj is active before applying it ...

Comment (Eric89GXL) [inline mne/tests/test_epochs.py]: > So, if preload=True the proj is applied a second time during get_data()

I think now it only is done that way if `add_eeg_ref=True` and there wasn't one before. But yes reapplying shouldn't matter (as long as all channels are present, [but that's a separate issue] (https://github.com/mne-tools/mne-python/issues/2310))

File: doc/whats_new.rst (hunk 3 above)
Comment (Eric89GXL) [inline doc/whats_new.rst]: Had a brief chat with @agramfort a bit ago and he agreed, so I'll make a PR for that once 0.11 is out

**Comment by Eric89GXL** [comment]
@jasmainak comments addressed, only last commit is new. Can you have one last look to make sure you're happy? If so I'll merge once the CIs come back happy.

File: mne/channels/tests/test_channels.py [hunk 7]
```diff
@@ -71,10 +71,12 @@ def test_set_channel_types():
     # Test change to illegal channel type
     mapping = {'EOG 061': 'xxx'}
     assert_raises(ValueError, raw.set_channel_types, mapping)
+    # Test changing type if in proj (avg eeg ref here)
+    mapping = {'EEG 060': 'eog', 'EEG 059': 'ecg', 'EOG 061': 'seeg'}
+    assert_raises(RuntimeError, raw.set_channel_types, mapping)
     # Test type change
-    raw2 = Raw(raw_fname)
+    raw2 = Raw(raw_fname, add_eeg_ref=False)
```
Comment (jasmainak) [inline mne/channels/tests/test_channels.py]: so will this throw an error if `proj` has not been applied but is added and the user is trying to change the channel type?

Comment (Eric89GXL) [inline mne/channels/tests/test_channels.py]: yeah

Comment (jasmainak) [inline mne/channels/tests/test_channels.py]: okay thanks for clarifying.

LGTM +1 for merge.
